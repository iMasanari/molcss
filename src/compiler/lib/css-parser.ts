import { AstEntity, AstRule, createParser } from 'css-selector-parser'
import { COMMENT, DECLARATION, Element, MEDIA, RULESET, SUPPORTS, compile } from 'stylis'

export const DUMMY_NESTING_SELECTOR_CLASS_NAME = '__MOLCSS_DUMMY_NESTING_SELECTOR_CLASS_NAME__'
export const DUMMY_NESTING_SELECTOR_CLASS_NAME_REG = new RegExp(DUMMY_NESTING_SELECTOR_CLASS_NAME, 'g')

const selectorParser = createParser()

const hasNestedSelector = (v: AstEntity) => v.type === 'ClassName' && v.name === DUMMY_NESTING_SELECTOR_CLASS_NAME
const hasPseudoElement = (v: AstEntity) => v.type === 'PseudoElement'

const isSelfSelector = (selector: string) => {
  const ast = selectorParser(selector)

  // 複数セレクタの場合、暫定的にfalse（実装めんどい）
  if (ast.rules.length !== 1) {
    return false
  }

  const rules: AstRule[] = []
  let target = ast.rules[0]

  while (target) {
    rules.push(target)
    target = target.nestedRule
  }

  const lastRule = rules[rules.length - 1]!

  // 最後に & が含まれている場合、疑似要素がなければ自分自身
  if (lastRule.items.some(hasNestedSelector)) {
    return !lastRule.items.some(hasPseudoElement)
  }

  // 全てに & がない場合、自分自身
  return !rules.some(v => v.items.some(hasNestedSelector))
}

export interface StyleData {
  group: string
  prop: string
  values: { selector: string, value: string, atRule: string[] }[]
}

export const parse = (style: string): StyleData[] => {
  const ast = compile(`.${DUMMY_NESTING_SELECTOR_CLASS_NAME}{${style}}`)
  const analyzed = ast.flatMap(element => analyzeStylisElement(element, [], ''))

  return [...new Set(analyzed.map(v => v.group))].map(group => {
    const list = analyzed.filter(v => v.group === group)
    const { prop } = list[0]!

    return {
      group,
      prop,
      values: list.map(({ selector, value, atRule }) => ({ selector, value, atRule })),
    }
  })
}

interface Analyzed {
  atRule: string[]
  selector: string
  prop: string
  value: string
  group: string
}

const analyzeStylisElement = (element: Element, meta: string[], selector: string): Analyzed[] => {
  switch (element.type) {
    case DECLARATION: {
      // prop: value;
      const prop = element.props as string
      const value = element.children as string
      selector = selector || DUMMY_NESTING_SELECTOR_CLASS_NAME

      const group = isSelfSelector(selector) ? prop : `${selector}{${prop}`

      return [{ group, atRule: meta, selector, prop, value }]
    }
    case RULESET: {
      // selector { ... }
      const children = element.children as Element[]

      return (element.props as string[]).flatMap(value => {
        const nextSelector = selector ? `${selector} ${value}` : value

        return children.flatMap(child => analyzeStylisElement(child, meta, nextSelector))
      })
    }
    case MEDIA:
    case SUPPORTS: {
      // @... { ... }
      const nextMeta = [...meta, element.value]
      const children = element.children as Element[]

      return children.flatMap(child => analyzeStylisElement(child, nextMeta, selector))
    }
    case COMMENT:
    default: {
      return []
    }
  }
}
