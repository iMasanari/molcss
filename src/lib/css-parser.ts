import { COMMENT, compile, DECLARATION, Element, MEDIA, RULESET } from 'stylis'

export interface StyleData {
  media: string
  selector: string
  prop: string
  values: string[]
}

export const parse = (style: string): StyleData[] => {
  const ast = compile(style)
  const analyzed = ast.flatMap(element => analyzeStylisElement(element, '', ''))

  return [...new Set(analyzed.map(v => v.group))].map(group => {
    const list = analyzed.filter(v => v.group === group)
    const { media, selector, prop } = list[0]!

    return {
      media,
      selector,
      prop,
      values: list.map(v => v.value),
    }
  })
}

interface Analyzed {
  media: string
  selector: string
  prop: string
  value: string
  group: string
}

const analyzeStylisElement = (element: Element, media: string, selector: string): Analyzed[] => {
  switch (element.type) {
    case DECLARATION: {
      // prop: value;
      const prop = element.props as string
      const value = element.children as string
      const group = `${media}{${selector || '&\f'}{${prop}`

      return [{ group, media, selector: selector || '&\f', prop, value }]
    }
    case COMMENT: {
      return []
    }
    case RULESET: {
      // selector { ... }
      const nextSelector = selector ? `${selector} ${element.value}` : element.value
      const children = element.children as Element[]

      return children.flatMap(child => analyzeStylisElement(child, media, nextSelector))
    }
    case MEDIA: {
      const nextMedia = element.value
      const children = element.children as Element[]

      return children.flatMap(child => analyzeStylisElement(child, nextMedia, selector))
    }
    default: {
      return []
    }
  }
}
