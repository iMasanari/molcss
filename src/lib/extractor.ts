const separator = ':::MOLCSS_SEPARATOR:::'

const contextTokens = [
  // tag
  '${', '}',
  // escape
  '\\',
  // literal
  `'`,
  `"`,
  '`',
  // comment
  '/*', '*/',
  '//', '\n',
] as const

type DefineToken = typeof contextTokens[number]

const escapeRegExp = (string: string) =>
  string.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&')

const createTokenRegExp = (tag: string) =>
  new RegExp(`(${[tag, ...contextTokens].map(v => escapeRegExp(v)).join('|')})`)

interface RootMode {
  type: 'Root'
}

interface ContentMode {
  type: 'Content'
  parent: RootMode
}
interface CodeMode {
  type: 'Code'
  parent: ContentMode | TemplateLiteralMode
}
interface StringLiteralMode {
  type: 'StringLiteral'
  parent: CodeMode | RootMode
  endToken: DefineToken
}
interface TemplateLiteralMode {
  type: 'TemplateLiteral'
  parent: CodeMode
}
interface CommentMode {
  type: 'Comment'
  parent: CodeMode | RootMode
  endToken: DefineToken
}
interface EscapeMode {
  type: 'Escape'
  parent: Mode
  display: boolean
}

type Mode =
  | ContentMode
  | CodeMode
  | StringLiteralMode
  | TemplateLiteralMode
  | CommentMode
  | RootMode
  | EscapeMode

const createContentMode = (parent: RootMode): ContentMode =>
  ({ type: 'Content', parent })

const createCodeMode = (parent: ContentMode | TemplateLiteralMode): CodeMode =>
  ({ type: 'Code', parent })

const createEscapeMode = (parent: Mode, display: boolean): EscapeMode =>
  ({ type: 'Escape', parent, display })

const createStringLiteralMode = (parent: CodeMode | RootMode, endToken: DefineToken): StringLiteralMode =>
  ({ type: 'StringLiteral', parent, endToken })

const createCommentMode = (parent: CodeMode | RootMode, endToken: DefineToken): CommentMode =>
  ({ type: 'Comment', parent, endToken })

const createTemplateLiteralMode = (parent: CodeMode): TemplateLiteralMode =>
  ({ type: 'TemplateLiteral', parent })

const contentReducer = (state: ContentMode, currentToken: string, restTokens: string[]): string | null => {
  switch (currentToken as DefineToken) {
    case '`': {
      return ''
    }
    case '\\': {
      return reducer(createEscapeMode(state, true), restTokens)
    }
    case '${': {
      const result = reducer(createCodeMode(state), restTokens)

      return result != null ? `\${${separator}}` + result : result
    }
    default: {
      const result = reducer(state, restTokens)

      return result != null ? currentToken + result : result
    }
  }
}

const templateLiteralReducer = (state: TemplateLiteralMode, currentToken: string, restTokens: string[]): string | null => {
  switch (currentToken as DefineToken) {
    case '`': {
      return reducer(state.parent, restTokens)
    }
    case '\\': {
      return reducer(createEscapeMode(state, false), restTokens)
    }
    case '${': {
      return reducer(createCodeMode(state), restTokens)
    }
    default: {
      return reducer(state, restTokens)
    }
  }
}

const codeReducer = (state: CodeMode, currentToken: string, restTokens: string[]): string | null => {
  switch (currentToken as DefineToken) {
    case '}': {
      return reducer(state.parent, restTokens)
    }
    case '\\': {
      return reducer(createEscapeMode(state, false), restTokens)
    }
    case '//': {
      return reducer(createCommentMode(state, '\n'), restTokens)
    }
    case '/*': {
      return reducer(createCommentMode(state, '*/'), restTokens)
    }
    case `'`: {
      return reducer(createStringLiteralMode(state, `'`), restTokens)
    }
    case `"`: {
      return reducer(createStringLiteralMode(state, `"`), restTokens)
    }
    case '`': {
      return reducer(createTemplateLiteralMode(state), restTokens)
    }
    default: {
      return reducer(state, restTokens)
    }
  }
}

const stringLiteralReducer = (state: StringLiteralMode, currentToken: string, restTokens: string[]): string | null => {
  switch (currentToken as DefineToken) {
    case state.endToken: {
      return reducer(state.parent, restTokens)
    }
    case '\\': {
      return reducer(createEscapeMode(state, false), restTokens)
    }
    default: {
      return reducer(state, restTokens)
    }
  }
}

const commentReducer = (state: CommentMode, currentToken: string, restTokens: string[]): string | null => {
  switch (currentToken as DefineToken) {
    case state.endToken: {
      return reducer(state.parent, restTokens)
    }
    default: {
      return reducer(state, restTokens)
    }
  }
}

const rootReducer = (state: RootMode, currentToken: string, restTokens: string[]): string | null => {
  switch (currentToken as DefineToken) {
    case '//': {
      return reducer(createStringLiteralMode(state, '\n'), restTokens)
    }
    case '/*': {
      return reducer(createStringLiteralMode(state, '*/'), restTokens)
    }
    case '`': {
      return reducer(createContentMode(state), restTokens)
    }
    default: {
      if (/^\s+$/.test(currentToken)) {
        return reducer(state, restTokens)
      }

      return null
    }
  }
}

const escapeReducer = (state: EscapeMode, currentToken: string, restTokens: string[]): string | null => {
  const result = reducer(state.parent, restTokens)

  return state.display ? currentToken + result : result
}

const reducer = (state: Mode, tokens: string[]): string | null => {
  const [currentToken, ...restTokens] = tokens

  if (currentToken == null) {
    return null
  }

  switch (state.type) {
    case 'Root': {
      return rootReducer(state, currentToken, restTokens)
    }
    case 'Content': {
      return contentReducer(state, currentToken, restTokens)
    }
    case 'Code': {
      return codeReducer(state, currentToken, restTokens)
    }
    case 'TemplateLiteral': {
      return templateLiteralReducer(state, currentToken, restTokens)
    }
    case 'StringLiteral': {
      return stringLiteralReducer(state, currentToken, restTokens)
    }
    case 'Comment': {
      return commentReducer(state, currentToken, restTokens)
    }
    case 'Escape': {
      return escapeReducer(state, currentToken, restTokens)
    }
  }
}

export const extract = (source: string, tag: string) => {
  const tokens = source.split(createTokenRegExp(tag)).filter(Boolean)

  const contetns = tokens.flatMap((v, i) => {
    if (v !== tag) {
      return []
    }

    const text = reducer({ type: 'Root' }, tokens.slice(i + 1))

    if (!text) {
      return []
    }

    return { quasis: text.split(`\${${separator}}`) }
  })

  return contetns
}
