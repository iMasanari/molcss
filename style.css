/* [molcss]: virtual module */
@layer molcss;