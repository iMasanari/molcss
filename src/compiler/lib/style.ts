import { defineProperties, shorthands } from '../../generated/defineCssProperties'
import { convertToAlphabet } from '../../utils/convertor'
import { DUMMY_NESTING_SELECTOR_CLASS_NAME_REG, StyleData } from './css-parser'

const hasOwn = Object.prototype.hasOwnProperty

interface StyleKeyData {
  group: string
  values: Map<string, string>
}

export type StyleContext = Map<string, StyleKeyData>

export const createStyleContext = (): StyleContext => {
  return new Map([...defineProperties.entries()].map(([prop, group]) =>
    [prop, { group, values: new Map() }]
  ))
}

const getOrCreate = <T, U>(collection: Map<T, U>, key: T, create: (collection: Map<T, U>) => U) => {
  if (collection.has(key)) {
    return collection.get(key)!
  }

  const value = create(collection)
  collection.set(key, value)

  return value
}

export const createClassName = (result: StyleData, context: StyleContext) => {
  const styleGroupData = getOrCreate(context, result.group, () => ({
    group: convertToAlphabet(context.size),
    values: new Map<string, string>(),
  }))

  const styleValueName = result.values.length
    ? getOrCreate(styleGroupData.values, result.values.map(v => [...v.atRule, , v.selector, v.value].join('{')).join('\n'), (v) => v.size.toString())
    : '00' // ランタイムスタイルのキー取得用

  return styleGroupData.group + styleValueName
}

export const createRuntimeKey = (styleData: StyleData, index: number, styleContext: StyleContext) => {
  const group = `__MOLCSS_RUNTIME_KEY__${styleData.group}__${index}`
  const className = createClassName({ prop: group, values: [], group }, styleContext)

  return className.replace(/\d+$/, '')
}

export const createStyle = (styleMap: Map<string, StyleData>) => {
  return [...styleMap.entries()].sort(sortFn).map(([selector, data]) => getStyle(data, selector)).join('\n') + '\n'
}

const isShorthand = (prop: string) => {
  const key = defineProperties.get(prop)

  if (!key) {
    return false
  }

  return hasOwn.call(shorthands, key)
}

const sortFn = ([, a]: [unknown, StyleData], [, b]: [unknown, StyleData]) => {
  const aIsShorthand = isShorthand(a.prop)

  if (aIsShorthand !== isShorthand(b.prop)) {
    return aIsShorthand ? -1 : 1
  }

  if (a.prop !== b.prop) {
    // Sort by vendor prefix first.
    return a.prop < b.prop ? -1 : 1
  }

  return a.group > b.group ? -1 : 1
}

const getStyle = (result: StyleData, selector: string) => {
  const styles = result.values.map(v => {
    const style = `${v.selector.replace(DUMMY_NESTING_SELECTOR_CLASS_NAME_REG, selector)}{${result.prop}:${v.value}}`

    return v.atRule.reduceRight((acc, v) => `${v}{${acc}}`, style)
  })

  return styles.join(' ')
}
