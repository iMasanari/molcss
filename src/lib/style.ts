import { defineProperties, shorthands } from '../generated/defineCssProperties'
import { convertToAlphabet } from '../utils/convertor'
import { StyleData } from './css-parser'

const hasOwn = Object.prototype.hasOwnProperty

interface StyleKeyData {
  name: string
  values: Map<string, string>
  meta: Map<string, string>
}

export type StyleContext = Map<string, StyleKeyData>

export const createStyleContext = (): StyleContext => {
  return new Map([...defineProperties.entries()].map(([prop, name]) =>
    [prop, { name, values: new Map(), meta: new Map() }]
  ))
}

const getOrCreate = <T, U>(collection: Map<T, U>, key: T, create: (collection: Map<T, U>) => U) => {
  if (collection.has(key)) {
    return collection.get(key)!
  }

  const value = create(collection)
  collection.set(key, value)

  return value
}

export const createClassName = (result: StyleData, context: StyleContext) => {
  const stylePropData = getOrCreate(context, result.prop, () => ({
    name: convertToAlphabet(context.size),
    values: new Map<string, string>(),
    meta: new Map<string, string>(),
  }))

  const styleValueName = result.values.length
    ? getOrCreate(stylePropData.values, result.values.join('\n'), (v) => v.size.toString())
    : '00' // ランタイムスタイルのキー取得用

  const styleMetaName = result.media || result.selector !== '&\f'
    ? getOrCreate(stylePropData.meta, `${result.media}{${result.selector}`, (v) => convertToAlphabet(v.size))
    : ''

  return stylePropData.name + styleValueName + styleMetaName
}

export const createRuntimeKey = (styleData: StyleData, styleContext: StyleContext) => {
  const token = createClassName({ ...styleData, values: [] }, styleContext)
  const className = createClassName({ prop: `--molcss-runtime-key-${token}`, values: [], selector: '&\f', media: '' }, styleContext)

  return className.replace(/\d+$/, '')
}

export const createStyle = (styleMap: Map<string, StyleData>) => {
  return [...styleMap.entries()].sort(sortFn).map(([className, data]) => getStyle(data, className)).join('\n') + '\n'
}

const isShorthand = (prop: string) => {
  const key = defineProperties.get(prop)

  if (!key) {
    return false
  }

  return hasOwn.call(shorthands, key)
}

const sortFn = ([, a]: [string, StyleData], [, b]: [string, StyleData]) => {
  if (a.media !== b.media) {
    if (!a.media) return -1
    if (!b.media) return 1

    return a.media > b.media ? -1 : 1
  }

  const aIsShorthand = isShorthand(a.prop)

  if (aIsShorthand !== isShorthand(b.prop)) {
    return aIsShorthand ? -1 : 1
  }

  if (a.prop !== b.prop) {
    // Sort by vendor prefix first.
    return a.prop < b.prop ? -1 : 1
  }

  return a.selector > b.selector ? -1 : 1
}

const getStyle = (result: StyleData, className: string) => {
  const selector = result.selector.replace(/&\f?/g, `.${className}`)

  const styles = result.values.map(v => {
    const style = `${selector}{${result.prop}:${v}}`

    return result.media ? `${result.media}{${style}}` : style
  })

  return styles.join('')
}
