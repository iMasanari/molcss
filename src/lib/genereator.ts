import { StyleData, parse } from './css-parser'
import { extract } from './extractor'

const cache = new Map<string, StyleData[]>()

export const generate = (source: string) => {
  const cssTexts = extract(source)

  const result = cssTexts.flatMap((css) => {
    const key = css.trim()
    const ref = cache.get(key)
    if (ref) return ref

    const styles = parse(css)
    cache.set(key, styles)

    return styles
  })

  return result
}
