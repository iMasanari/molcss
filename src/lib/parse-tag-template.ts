import { parse } from './css-parser'
import { StyleContext, createRuntimeKey } from './style'

interface RuntimeStyleData<T> {
  prop: string
  expressions: (string | T)[]
}

export const parseTagTemplate = <T>(quasis: string[], expressions: T[], styleContext: StyleContext) => {
  let styleText = ''
  const expressionMap = new Map<string, T>()
  const lastIndex = quasis.length - 1

  for (const [i, v] of quasis.entries()) {
    styleText += v

    if (i !== lastIndex) {
      const key = `__MOLCSS_VALUE_${i}__`
      styleText += key
      expressionMap.set(key, expressions[i]!)
    }
  }

  const styles = parse(styleText)

  const runtimeStyles: RuntimeStyleData<T>[] = []

  for (const styleData of styles) {
    for (const [index, { value }] of styleData.values.entries()) {
      const list = value.split(/(__MOLCSS_VALUE_\d+__)/)

      if (list.length < 2) {
        continue
      }

      const classPropKey = createRuntimeKey(styleData, index, styleContext)

      styleData.values[index]!.value = `var(--molcss-${classPropKey})`

      runtimeStyles.push({
        prop: classPropKey,
        expressions: list.filter(Boolean).map(v =>
          expressionMap.has(v) ? expressionMap.get(v)! : v
        ),
      })
    }
  }

  return { styles, runtimeStyles }
}
