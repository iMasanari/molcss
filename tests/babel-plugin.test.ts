import { BabelFileResult, transformAsync } from '@babel/core'
import { expect, it } from 'vitest'
import plugin from '../src/babel-plugin'
import { createStyle } from '../src/lib/style'

const createStyleFromActual = (actual: BabelFileResult) => {
  const styleData = (actual.metadata as any).molcss

  return createStyle(styleData)
}

it('css`...`', async () => {
  const code = `
    import { css } from "molcss";
    
    css\`
      color: red;
    \`;
  `

  const actual = await transformAsync(code, {
    plugins: [[plugin, { devLabel: false }]],
  })

  expect(actual?.code).toMatchInlineSnapshot('"\\"c0\\";"')
  expect(createStyleFromActual(actual!)).toMatchInlineSnapshot(`
    ".c0{color:red}
    "
  `)
})

it('css`${...}`', async () => {
  const code = `
    import { css } from "molcss";
    
    const cssColor = 'blue';
    
    css\`
      color: \${cssColor};
      border: 1px solid \${cssColor};
    \`;
  `

  const actual = await transformAsync(code, {
    plugins: [[plugin, { devLabel: false }]],
  })

  expect(actual?.code).toMatchInlineSnapshot(`
    "const cssColor = 'blue';
    ({
      className: \\"c0 m0\\",
      runtime: [[\\"bL\\", cssColor], [\\"bM\\", \\"1px solid \\" + cssColor]]
    });"
  `)

  expect(createStyleFromActual(actual!)).toMatchInlineSnapshot(`
    ".m0{border:var(--molcss-bM)}
    .c0{color:var(--molcss-bL)}
    "
  `)
})

it('devLabel option (identifier)', async () => {
  const code = `
    import { css } from "molcss";
    
    const identifierStyle = css\`
      color: red;
    \`;
  `

  const actual = await transformAsync(code, {
    filename: 'devLabel option (identifier)',
    plugins: [[plugin, { devLabel: true }]],
  })

  expect(actual?.code).toMatchInlineSnapshot(`
    "const identifierStyle = \\"DEV-devLabelOptionIdentifier-identifierStyle c0\\";"
  `)

  expect(createStyleFromActual(actual!)).toMatchInlineSnapshot(`
    ".c0{color:red}
    "
  `)
})

it('devLabel option (object)', async () => {
  const code = `
    import { css } from "molcss";

    const objectStyle = {
      identifierStyle: css\`
        color: red;
      \`,
      'literalStyle': css\`
        color: green;
      \`,
    }
  `

  const actual = await transformAsync(code, {
    filename: 'devLabel option (object)',
    plugins: [[plugin, { devLabel: true }]],
  })

  expect(actual?.code).toMatchInlineSnapshot(`
    "const objectStyle = {
      identifierStyle: \\"DEV-devLabelOptionObject-identifierStyle c0\\",
      'literalStyle': \\"DEV-devLabelOptionObject-literalStyle c1\\"
    };"
  `)

  expect(createStyleFromActual(actual!)).toMatchInlineSnapshot(`
    ".c0{color:red}
    .c1{color:green}
    "
  `)
})

it('devLabel option (function)', async () => {
  const code = `
    import { css } from "molcss";

    function functionDeclaration() {
      css\`
        color: red;
      \`
    }

    const _ = function functionExpression() {
      css\`
        color: green;
      \`
    }

    const variableDeclaration = function () {
      css\`
        color: blue;
      \`
    }
  `

  const actual = await transformAsync(code, {
    filename: 'devLabel option (function).js',
    plugins: [[plugin, { devLabel: true }]],
  })

  expect(actual?.code).toMatchInlineSnapshot(`
    "function functionDeclaration() {
      \\"DEV-devLabelOptionFunction-functionDeclaration c0\\";
    }
    const _ = function functionExpression() {
      \\"DEV-devLabelOptionFunction-functionExpression c1\\";
    };
    const variableDeclaration = function () {
      \\"DEV-devLabelOptionFunction-variableDeclaration c2\\";
    };"
  `)

  expect(createStyleFromActual(actual!)).toMatchInlineSnapshot(`
    ".c0{color:red}
    .c1{color:green}
    .c2{color:blue}
    "
  `)
})

it('devLabel option (class)', async () => {
  const code = `
    import { css } from "molcss";

    class ClassDeclaration {
      render() {
        css\`
          color: red;
        \`
      }
    }

    const _ = class ClassExpression {
      render() {
        css\`
          color: green;
        \`
      }
    }

    const VariableDeclaration = class {
      render() {
        css\`
          color: blue;
        \`
      }
    }
  `

  const actual = await transformAsync(code, {
    filename: 'devLabel option (class).js',
    plugins: [[plugin, { devLabel: true }]],
  })

  expect(actual?.code).toMatchInlineSnapshot(`
    "class ClassDeclaration {
      render() {
        \\"DEV-devLabelOptionClass-ClassDeclaration c0\\";
      }
    }
    const _ = class ClassExpression {
      render() {
        \\"DEV-devLabelOptionClass-ClassExpression c1\\";
      }
    };
    const VariableDeclaration = class {
      render() {
        \\"DEV-devLabelOptionClass-VariableDeclaration c2\\";
      }
    };"
  `)

  expect(createStyleFromActual(actual!)).toMatchInlineSnapshot(`
    ".c0{color:red}
    .c1{color:green}
    .c2{color:blue}
    "
  `)
})
