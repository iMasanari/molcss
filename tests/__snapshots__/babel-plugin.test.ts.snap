// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`css\`\${...}\` 1`] = `
"const cssColor = 'blue';
({
  className: \\"c0 m0\\",
  runtime: [[\\"bL\\", cssColor], [\\"bM\\", \\"1px solid \\" + cssColor]]
});"
`;

exports[`css\`...\` 1`] = `
[
  "c0",
]
`;

exports[`css\`...\` 2`] = `"\\"c0\\";"`;
